File: 983_minimum_cost_for_tickets.java
class Solution {
    // Time Complexity: O(365)
    // Space Complexity: O(365)
    public int mincostTickets(int[] days, int[] costs) {
        Integer[] memo = new Integer[366];
        Set<Integer> dayset = new HashSet<>();
        for (int d: days) dayset.add(d);

        return dp(1, memo, dayset, costs);
    }

    public int dp(int day, Integer[] memo, Set<Integer> dayset, int[] costs) {
        if (day > 365) {
            return 0;
        }

        if (memo[day] != null) {
            return memo[day];
        }

        int ans;
        if (dayset.contains(day)) {
            ans = Math.min(dp(day + 1, memo, dayset, costs) + costs[0], dp(day + 7, memo, dayset, costs) + costs[1]);
            ans = Math.min(ans, dp(day + 30, memo, dayset, costs) + costs[2]);
        } else {
            ans = dp(day + 1, memo, dayset, costs);
        }

        memo[day] = ans;
        return ans;
    }
}

File: python/983_minimum_cost_for_tickets.py
class Solution:
    def mincostTickets(self, days: List[int], costs: List[int]) -> int:
        arr = [None] * 366
        day_set = set(days)

        def dp(i: int, arr: list, day_set: set) -> int:
            if i > 365:
                return 0
            if arr[i]:
                return arr[i]
            ans = -1
            if i in day_set:
                ans = min(dp(i + 1, arr, day_set) + costs[0],   \
                          dp(i + 7, arr, day_set) + costs[1], dp(i + 30, arr, day_set) + costs[2])
            else:
                ans = dp(i + 1, arr, day_set)

            arr[i] = ans
            return ans

        return dp(1, arr, day_set)
